### PR TITLE
GlobalSign DSS client API

### PR DESCRIPTION
**Pull request by sampila**

### CODE REVIEW

File: certificate_path.go [hunk 1]
```diff
@@ -0,0 +1,6 @@
+package globalsign
```
Comment (gunnsth) [inline certificate_path.go:1]: https://github.com/unidoc/unipdf/wiki/UniPDF-Developer-Guide#8-license-header-in-each-go-file
Need to add a license too... Probably we can distribute this as MIT, but maybe need to make a note that it links to UniPDF which has a higher license requirement, i.e. EULA?

File: certificate_path.go [hunk 2]
```diff
@@ -0,0 +1,6 @@
...
+
+// CertificatePathResponse .
```
Comment (gunnsth) [inline certificate_path.go]: This comment is not really appropriate, it should be a full sentence, go doc style.
Is it necessary to export this? I.e. is it used outside?

File: client.go [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"sync"
+	"time"
+)
+
+// Client GlobalSign sdk client.
```
Comment (gunnsth) [inline client.go]: This is a broken sentence, should improve it... Is this also used externally?
We can unexport anything not needed externally

File: client.go [hunk 4]
```diff
@@ -0,0 +1,277 @@
+package globalsign
```
Comment (gunnsth) [inline client.go:1]: header also

File: client.go [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+	ApiSecret string
+
+	// CertFilePath path file to mTLS cert file.
+	CertFilePath string
+
+	// KeyFilePath path file to mTLS cert private key.
+	KeyFilePath string
+}
+
+// NewClient initiate client with `API Key`, `API Secret`, `Certificate file path`, `Private Key file path` and return globalsign sdk client.
```
Comment (gunnsth) [inline client.go]: Should be "initiates" and then "returns the globalsign client"

File: client.go [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
+			tls.TLS_RSA_WITH_AES_256_CBC_SHA,
+		},
+		Certificates: []tls.Certificate{cert},
+	}
+	tr := &http.Transport{TLSClientConfig: config}
+
+	return &http.Client{Transport: tr}, nil
+}
+
+func (c *Client) DoNewRequest(method, path string, result interface{}, params interface{}) error {
```
Comment (gunnsth) [inline client.go:126]: missing godoc, is this also needed as exported?

File: client.go [hunk 7]
```diff
@@ -0,0 +1,277 @@
...
+	err = c.Do(req, result)
+	if err != nil {
+		return err
+	}
+
+	// Return no errors on success.
+	return nil
+}
+
+func (c *Client) NewRequest(method string, path string, params interface{}) (*http.Request, error) {
```
Comment (gunnsth) [inline client.go:144]: also

File: client.go [hunk 8]
```diff
@@ -0,0 +1,277 @@
...
+		if err != nil {
+			return nil, err
+		}
+	}
+	req.Header.Add("Content-Type", contentType)
+
+	return req, nil
+}
+
+func (c *Client) Do(req *http.Request, result interface{}) error {
```
Comment (gunnsth) [inline client.go:192]: also

File: client.go [hunk 9]
```diff
@@ -0,0 +1,277 @@
...
+	if err != nil {
+		return err
+	}
+
+	// Unmarshal the response.
+	err = json.Unmarshal(data, result)
+	return err
+}
+
+// SetAuthToken set authentication token to client.
```
Comment (gunnsth) [inline client.go]: SetAuthToken sets the authentication token of the client.

File: dss.go [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/hex"
+	"strings"
+	"time"
+)
+
+// DSSService .
```
Comment (gunnsth) [inline dss.go]: broken comment...
Do we need this interface?  Where is it implemented?

File: dss.go [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+	Sign(*SigningRequest) (*SigningResponse, error)
+	CertificatePath() (*CertificatePathResponse, error)
+	TrustChain() (*TrustChainResponse, error)
+	// DSS Identity and sign process.
+	DSSGetIdentity(context.Context, string, *IdentityRequest) (*DSSIdentity, error)
+	DSSIdentitySign(context.Context, string, *IdentityRequest, []byte) ([]byte, error)
+	DSSIdentityTimestamp(context.Context, string, *IdentityRequest, []byte) ([]byte, error)
+}
+
+// DSSIdentity represent acquired credential
```
Comment (gunnsth) [inline dss.go]: represents, maybe fix the line length etc?

File: dss.go [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+// from login and identity request.
+type DSSIdentity struct {
+	// Identity.
+	ID string
+
+	// SigningCert.
```
Comment (gunnsth) [inline dss.go]: Kind of pointless comment... if there is no further information

File: dss.go [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+	OCSP string
+
+	//CA Certificate.
+	CA string
+
+	// Ts timestamp.
+	Ts time.Time
+}
+
+// GetIdentity .
```
Comment (gunnsth) [inline dss.go]: broken comment

File: dss.go [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+		SigningCert: identityResp.SigningCert,
+		OCSP:        identityResp.OCSPResponse,
+		CA:          certResp.CA,
+	}
+	s.client.vault.Set(signer, identity)
+
+	return identity, nil
+}
+
+// Sign .
```
Comment (gunnsth) [inline dss.go]: bad comment

File: dss.go [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+		Digest: digestHex,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return hex.DecodeString(signatureResp.Signature)
+}
+
+// Timestamp .
```
Comment (gunnsth) [inline dss.go]: check

File: identity.go [hunk 16]
```diff
@@ -0,0 +1,34 @@
+package globalsign
+
+// IdentityRequest .
+type IdentityRequest struct {
```
Comment (gunnsth) [inline identity.go:8]: is this needed externally?

File: identity.go [hunk 17]
```diff
@@ -0,0 +1,34 @@
+package globalsign
+
+// IdentityRequest .
```
Comment (gunnsth) [inline identity.go]: bad comment

File: identity.go [hunk 18]
```diff
@@ -0,0 +1,34 @@
+package globalsign
```
Comment (gunnsth) [inline identity.go:1]: header

File: timestamp.go [hunk 19]
```diff
@@ -0,0 +1,11 @@
+package globalsign
+
+// TimestampRequest .
+type TimestampRequest struct {
+	Digest string `json:"digest"`
+}
+
+// TimestampResponse .
```
Comment (gunnsth) [inline timestamp.go]: more broken comments

**Review by gunnsth — CHANGES_REQUESTED** [review]
Did one pass of review and added a few comments to check.  Good to apply the comments throughout the package.

File: certificate_path.go (hunk 2 above)
Comment (sampila) [reply certificate_path.go]: The point of making request and response struct exported is because, we make client API also, so users can do the API request to GlobalSign service if needed.

File: client.go (hunk 9 above)
Comment (sampila) [reply client.go]: Updated at c601d01cd3b73431e6ad8a2d86486c05ef75bcc1

File: dss.go (hunk 11 above)
Comment (sampila) [reply dss.go]: Updated at a3ddc191292d0952b56e2269c3cbfa6f517b98fa

File: dss.go (hunk 12 above)
Comment (sampila) [reply dss.go]: Updated at a3ddc191292d0952b56e2269c3cbfa6f517b98fa

File: dss.go (hunk 13 above)
Comment (sampila) [reply dss.go]: Updated at a3ddc191292d0952b56e2269c3cbfa6f517b98fa

File: dss.go (hunk 14 above)
Comment (sampila) [reply dss.go]: Updated at a3ddc191292d0952b56e2269c3cbfa6f517b98fa

File: dss.go (hunk 15 above)
Comment (sampila) [reply dss.go]: Updated at a3ddc191292d0952b56e2269c3cbfa6f517b98fa

File: identity.go (hunk 17 above)
Comment (sampila) [reply identity.go]: Updated at a3ddc191292d0952b56e2269c3cbfa6f517b98fa

File: timestamp.go (hunk 19 above)
Comment (sampila) [reply timestamp.go]: Updated at a3ddc191292d0952b56e2269c3cbfa6f517b98fa

File: certificate_path.go (hunk 1 above)
Comment (sampila) [reply certificate_path.go:1]: Updated at 1e197714ba820fb2d596cd1880cdf67ff93abed0

File: client.go (hunk 8 above)
Comment (sampila) [reply client.go:192]: Updated at a3ddc191292d0952b56e2269c3cbfa6f517b98fa

File: client.go (hunk 5 above)
Comment (sampila) [reply client.go]: Updated at a3ddc191292d0952b56e2269c3cbfa6f517b98fa

File: dss.go (hunk 10 above)
Comment (sampila) [reply dss.go]: Updated at a3ddc191292d0952b56e2269c3cbfa6f517b98fa

File: identity.go (hunk 18 above)
Comment (sampila) [reply identity.go:1]: About license updated at 1e197714ba820fb2d596cd1880cdf67ff93abed0

File: client.go [hunk 20]
```diff
@@ -0,0 +1,280 @@
...
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"sync"
+	"time"
+)
+
+// Client globasign client.
```
Comment (gunnsth) [inline client.go]: // Client implements a client for Globalsign Atlas DSS API service.

File: client.go [hunk 21]
```diff
@@ -0,0 +1,280 @@
...
+	ApiSecret string
+
+	// CertFilePath path file to mTLS cert file.
+	CertFilePath string
+
+	// KeyFilePath path file to mTLS cert private key.
+	KeyFilePath string
+}
+
+// NewClient initiates client with `API Key`, `API Secret`, `Certificate file path`, `Private Key file path` and returns the globalsign client.
```
Comment (gunnsth) [inline client.go]: normally only put `variable` names inside ticks.  Can just skip the ticks I think for this

File: dss.go [hunk 22]
```diff
@@ -0,0 +1,119 @@
...
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/hex"
+	"strings"
+	"time"
+)
+
+// DSSService availables GlobalSign Digital Signing Service.
```
Comment (gunnsth) [inline dss.go]: availables?  Not sure what that is supposed to mean.. Maybe "implements" or "defines what is needed for..."?

File: quotas.go [hunk 23]
```diff
@@ -0,0 +1,36 @@
+package globalsign
+
+import (
+	"net/http"
+)
+
+// QuotasResponse GlobalSign API response of `/quotas/signatures` and `/quotas/timestamps` endpoint.
+type QuotasResponse struct {
+	Value int `json:"value"`
```
Comment (gunnsth) [inline quotas.go:9]: does this have any units? signatures per some time period?

**Review by gunnsth — COMMENTED** [review]
Looks great, few minor comments

File: quotas.go (hunk 23 above)
Comment (sampila) [reply quotas.go:9]: Not having a unit for this one, only return a number of quotas as an integer.

File: client.go (hunk 20 above)
Comment (sampila) [reply client.go]: Updated at 6fa006e290519f3ae08c512c53671a7b8ce4e71f

File: client.go (hunk 21 above)
Comment (sampila) [reply client.go]: Updated at 6fa006e290519f3ae08c512c53671a7b8ce4e71f

File: dss.go (hunk 22 above)
Comment (sampila) [reply dss.go]: Updated at 7c0ab2d4eb0ae334b227906f9b4ba7513d876d31

File: client.go [hunk 24]
```diff
@@ -0,0 +1,281 @@
...
+	}
+	defer resp.Body.Close()
+
+	// Check the response for error.
+	if err := checkResponse(resp); err != nil {
+		return err
+	}
+
+	// Read all the data from the response.
+	data, err := ioutil.ReadAll(resp.Body)
```
Comment (adrg) [inline client.go]: Should be able to use `json.Decoder`. Something like:
```go
return json.NewDecoder(resp.Body).Decode(result)
```

File: dss.go [hunk 25]
```diff
@@ -0,0 +1,119 @@
...
+// DSSIdentity represent acquired credential from login and identity request.
+type DSSIdentity struct {
+	ID          string
+	SigningCert string
+	OCSP        string
+	CA          string
+	Ts          time.Time
+}
+
+// DSSGetIdentity get identity information, signing certificate, OCSP, and CA certificat -
```
Comment (adrg) [inline dss.go]: `CA certificate` instead of `CA certificat`.

File: trustchain.go [hunk 26]
```diff
@@ -0,0 +1,24 @@
+package globalsign
+
+import (
+	"net/http"
+)
+
+// TrusChainResponse GlobalSign API response of `/trustchain` endpoint.
```
Comment (adrg) [inline trustchain.go]: `TrustChainResponse` instead of `TrusChainResponse` in the comment.

File: trustchain.go [hunk 27]
```diff
@@ -0,0 +1,24 @@
...
+type TrustChainResponse struct {
+	Trustchain         []string `json:"trustchain"`
+	OcspRevocationInfo []string `json:"ocsp_revocation_info"`
+}
+
+// TrusChain GlobalSign DSS trustchain API service.
```
Comment (adrg) [inline trustchain.go]: `TrustChain` instead of `TrusChain` in the comment.
Might be good to run `golint` the entire package.

File: vault.go [hunk 28]
```diff
@@ -0,0 +1,136 @@
...
+		for {
+			select {
+			case <-ticker:
+				cache.cleanup()
+			}
+		}
+	})()
+}
+
+// NewIdentityVault is a helper to create instance of the indetities vault struct.
```
Comment (adrg) [inline vault.go]: `identities` instead of `indetities`.

File: dss.go [hunk 29]
```diff
@@ -0,0 +1,119 @@
...
+		return nil, err
+	}
+
+	// Request id and signing certificate.
+	identityResp, err := s.client.DSSService.Identity(req)
+	if err != nil {
+		return nil, err
+	}
+
+	// Request cs certificate.
```
Comment (adrg) [inline dss.go]: Should the comment say `Request CA certificate.`?

File: signing.go [hunk 30]
```diff
@@ -0,0 +1,35 @@
...
+
+import (
+	"net/http"
+)
+
+// SigningRequest GlobalSign API request parameters of `/identity/{id}/sign/{digest}` endpoint.
+type SigningRequest struct {
+	ID string `json:"id"`
+
+	// a hex encoded sha256 checksum for source file
```
Comment (adrg) [inline signing.go]: Can you rephrase this comment a bit?
`Digest is the hex encoded SHA256 checksum of a source file.`

**Review by adrg — COMMENTED** [review]
Looks good overall. I left a couple of minor observations.

File: client.go (hunk 24 above)
Comment (sampila) [reply client.go]: Updated at 04346b6b8dd2dacc90a70796c438dd47a60145a8

File: dss.go (hunk 25 above)
Comment (sampila) [reply dss.go]: Updated at 8cf21d161cd9788f60d69fcbce65d71b99707caa

File: trustchain.go (hunk 26 above)
Comment (sampila) [reply trustchain.go]: Updated at 8cf21d161cd9788f60d69fcbce65d71b99707caa

File: vault.go (hunk 28 above)
Comment (sampila) [reply vault.go]: Updated at 8cf21d161cd9788f60d69fcbce65d71b99707caa

File: dss.go (hunk 29 above)
Comment (sampila) [reply dss.go]: Updated at 8cf21d161cd9788f60d69fcbce65d71b99707caa

File: signing.go (hunk 30 above)
Comment (sampila) [reply signing.go]: Updated at 8cf21d161cd9788f60d69fcbce65d71b99707caa

File: trustchain.go (hunk 27 above)
Comment (sampila) [reply trustchain.go]: Updated at 45575c8e9becc74801fb68c4a32fb7dd82a5c8c4

**Review by adrg — APPROVED** [review]
Looks good. Thank you for addressing the review points.

**Review by gunnsth — APPROVED** [review]
LGTM
